ORC-189. Add timestamp with local timezone.

omalley opened the pull request.

prasanthj commented on java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java:
```diff
@@ -1634,6 +1634,30 @@ public int hashCode() {
+
+    @Override
+    public void updateTimestamp(Timestamp value) {
+      updateTimestamp(value.getTime());
```
do we always return the timestamp stats in local timezone? If so PPD on timestamp will still be broken right?

wgtmac replied:
I think this is correct. We will always compare the timestamp instant for the new type.

prasanthj commented on java/core/src/java/org/apache/orc/impl/writer/TimestampInstantTreeWriter.java:
```diff
@@ -0,0 +1,160 @@
+        if (millis < 0 && newNanos > 999_999) {
+          millis -= MILLIS_PER_SECOND;
+        }
+        indexStatistics.updateTimestamp(millis);
```
there seems to be some code duplication here. except for millis and seconds rest all looks the same between both timestamp writers. should the common code be factored into a common method? in case of any fixes will be easier to fix in one place. similarly for the readers.

prasanthj commented on java/core/src/java/org/apache/orc/impl/writer/TimestampInstantTreeWriter.java:
```diff
@@ -0,0 +1,160 @@
+      recordPosition(rowIndexPosition);
+    }
+    DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+    dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
```
nit: UTC timezone can be static somewhere. repeated in multiple places.

prasanthj commented on java/core/src/java/org/apache/orc/impl/ConvertTreeReaderFactory.java:
```diff
@@ -2886,6 +2898,7 @@ public static boolean canConvert(TypeDescription fileType, TypeDescription reade
       case LONG:
       case DOUBLE:
       case TIMESTAMP:
+      case TIMESTAMP_INSTANT:
```
conversion between TIMESTAMP and TIMESTAMP_INSTANT seems to be missing?

jcamachor commented:
@omalley , @prasanthj , would it be possible to get this in? Then we can implement the missing part in Hive to support instant semantics in ORC.

omalley commented:
Ok, I rewrote this patch.

Changes:
* The timestamp TreeReader/TreeWriter use the same code with a boolean on the constructor.
* Fixed the SchemaEvolution to work with the instant timestamps.
* Added a set of test cases to test the schema evolution. That brought up another set of problems:
  * The schema evolution wasn't being tested and wasn't documented.
  * Some of the conversions were very confusing.
  * I changed the previous semantics for timestamp -> long to use seconds instead of millis from 1970. This was based on the fact that timestamp -> decimal and double used seconds.
  * The conversions to timestamp were ignoring the option to useUtc.
  * The conversion to/from decimal were going through double and string, which loses precision.
  * The float conversions exactly mirror the double conversions, so I removed the timestamp ones rather than fixing them too. I created a new jira to refactor all of them.
* instant timestamp <-> string uses 'yyyy-mm-dd hh:mm:ss.s zone' where zone is the zone id. 
* statistics now depends on the type of the column, which took a little refactoring

prasanthj commented on java/core/src/java/org/apache/orc/impl/ConvertTreeReaderFactory.java:
```diff
@@ -1995,31 +1520,33 @@ public void nextVector(ColumnVector previousVector,
-          DateWritable.daysToMillis((int) longColVector.vector[elementNum]);
-      timestampColVector.set(elementNum, new Timestamp(millis));
+      long days = longColVector.vector[elementNum];
+      long millis = days * 24 * 60 * 60 * 1000;
```
"L" to avoid millis becoming negative?

prasanthj commented on java/core/src/java/org/apache/orc/impl/ConvertTreeReaderFactory.java:
```diff
@@ -1833,70 +1380,38 @@ public void nextVector(ColumnVector previousVector,
+      long wholeSec = (long) Math.floor(seconds);
+      timestampColVector.time[elementNum] = wholeSec * 1000;
+      timestampColVector.nanos[elementNum] =
+          1_000_000 * (int) Math.round((seconds - wholeSec) * 1000);
```
similarly add "L" to avoid int overflows.
